packages/templates/src/utils/memoize.ts

```diff
@@ -13,9 +13,14 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-export default function memoize(func, argsHash = (args) => args.join('-')) {
+
+type ArgHash = (args: string[] | number[]) => string;
+export default function memoize<T>(
+  func: (...args: string[]) => T,
+  argsHash: ArgHash = (args) => args.join('-')
+) {
   const memoized = new Map();
```

Comment:
Wait, `map.has()` works fine for `undefined`, so the whole `IS_VOID` nonsense is void (😎). I also cleaned up the arg stuff (why only all strings or all numbers?).

Here's a more condensed version (annoyingly the last cast is required, because TypeScript can't update it's expected type from the `.get()` based on the `.has()`):

```ts
type Arg = string | number;
type ArgHash = (args: Arg[]) => string;
export default function memoize<T>(
  func: (...args: Arg[]) => T,
  argsHash: ArgHash = (args) => args.join('-')
) {
  const memoized = new Map<string, T>();
  return function (...args: Arg[]) {
    const key = argsHash(args);
    if (!memoized.has(key)) {
      const result = func(...args);
      memoized.set(key, result);
      return result;
    }
    // Here get() has signature `T | undefined` but because of has() above,
    // it must be T (which can technically still include undefined, but that's
    // besides the point).
    return memoized.get(key) as T;
  };
}

```